torch/nn/parallel/_functions.py

```diff
@@ -51,7 +51,8 @@ class Gather(Function):
 
     @staticmethod
     def forward(ctx, target_device, dim, *inputs):
-        assert all(map(lambda i: i.is_cuda, inputs))
+        if any([input.device.type == "cpu" for input in inputs]):
+            raise TypeError('Gather function not implemented for CPU tensors')
```

Comment:
Shall preserve the behavior that this throws an `AssertError`? E.g., 

```python
assert all(map(lambda i: i.device.type == 'cpu', inputs)), (
    'Gather function not implemented for CPU tensors'
)
```

Reply:
`AssertError` is better.